system/jlib/jlzw.cpp

```diff
@@ -758,82 +758,76 @@ size32_t RLEExpand(void *dst,const void *src,size32_t expsize)
     return (size32_t)(in-(const byte *)src);
 }
 
-void appendToBuffer(MemoryBuffer & out, size32_t len, const void * src)
+void compressToBuffer(MemoryBuffer & out, size32_t len, const void * src, CompressionMethod method, const char *options)
 {
-    out.append(false);
-    out.append(len);
-    out.append(len, src);
-}
-
-void compressToBuffer(MemoryBuffer & out, size32_t len, const void * src)
-{
-    unsigned originalLength = out.length();
-    out.append(true);
-    out.append((size32_t)0);
-
-    if (len >= 32)
+    if (method != COMPRESS_METHOD_NONE && len >= 32)
     {
+        ICompressHandler *handler = queryCompressHandler(method);
+        if (!handler)
+        {
+            VStringBuffer s("Unknown compression method %x requested in compressToBuffer", (byte) method);
+            throw makeStringException(0, s.str());
+        }
+        unsigned originalLength = out.length();
+        // For back-compatibility, we always store COMPRESS_METHOD_LZW as 1 as earlier versions stored a boolean here
+        // rather than an enum
+        // This means that compressToBuffer/decompressToBuffer cannot bs used for rowdiff compression - this is not likely to be an issue
+        // Alternative would be a separate enum for compressToBuffer formats, but that seems more likely to cause confusion
+        out.append((byte) (method == COMPRESS_METHOD_LZW ? COMPRESS_METHOD_LZWLEGACY : method));
+        out.append((size32_t)0);
         size32_t newSize = len * 4 / 5; // Copy if compresses less than 80% ...
-        Owned<ICompressor> compressor = createLZWCompressor();
+        Owned<ICompressor> compressor = handler->getCompressor(options);
         void *newData = out.reserve(newSize);
-        compressor->open(newData, newSize);
-        if (compressor->write(src, len)==len)
+        try
         {
-            compressor->close();
-            size32_t compressedLen = compressor->buflen();
-            out.setWritePos(originalLength + sizeof(bool));
-            out.append(compressedLen);
-            out.setWritePos(originalLength + sizeof(bool) + sizeof(size32_t) + compressedLen);
-            return;
+            compressor->open(newData, newSize);
+            if (compressor->write(src, len)==len)
+            {
+                compressor->close();
+                size32_t compressedLen = compressor->buflen();
+                out.setWritePos(originalLength + sizeof(byte));
+                out.append(compressedLen);
```

Comment:
trivial: could use a DelayedMarker<size32_t> instead of line 777, and a marker.write(compressedLen) here